File: linked_lists/rmdup.java
import java.util.Arrays;


public class rmdup {

    public class List<V> {
        public Node<V> head;

        public List() {
            this.head = null;
        }

        public List<V> add(V value) {
            if (this.head == null) {
                this.head = new Node(value);
            } else {
                this.head.add(value);
            }
            return this;
        }

        public int size() {
            int i = 0;
            for (Node<V> n = this.head; n != null; n = n.next) {
                i++;
            }
            return i;
        }

        public V[] toArray() {
            final int n = this.size();
            V[] result = new V[n];

            Node<V> node = this.head;
            for (int i = 0; i < n; i++, node = node.next) {
                result[i] = node.value;
            }

            return result;
        }


    public class Node<V> {
        public Node<V> next;
        public V value;

        public Node(final V value) {
            this.value = value;
            this.next = null;
        }

        public void add(V value) {
            if (this.next == null) {
                this.next = new Node<V>(value);
            } else {
                this.next.add(value);
            }
        }
    }

    }


    public static String debugString(Integer[] i) {
        int[] result = new int[i.length];
        for (int j = 0; j < i.length; j++) {
            result[j] = i[j].intValue();
        }

        return Arrays.toString(result);
    }


    public static void test(List<Integer> input, Integer[] expected) {
        Integer[] actual = rmdup(input).toArray();
        if (Arrays.equals(expected, actual)) {
            System.out.printf("[SUCCESS] Test passed for input: %s\n", debugString(input));
        } else {
            System.out.printf("[FAIL]    Test failed for input: %s\n", debugString(input));
        }

        System.out.printf(" * Expected: %s\n", debugString(expected));
        System.out.printf(" * Actual:   %s\n", debugString(actual));
        System.out.println();
    }


    public static void main(String[] args) {
        Node n1 = new List<Integer>()
            .add(1)
            .add(2)
            .add(3);
        test(n1, new Integer[] { 1, 2, 3 });

        Node n2 = new List<Integer>()
            .add(1)
            .add(1)
            .add(1);
        test(n2, new Integer[] { 1 });
    }

}

File: ctci/linked_lists/rmdup/rmdup.c
#include<stdio.h>
#include<stdlib.h>
#include<string.h>


#define DEBUG 0


/**
 * Allocate new string of length n + 1, and
 * fill contents with 0's
 * TODO: Add error-checking for malloc call
 */
char * str_alloc(const int n) {
  char * s = malloc(sizeof(char) * n + 1);
  for (int i = 0; i < n; i++) {
	s[i] = '\0';
  }
  s[n] = '\0';
  return s;
}


/**
 * Shift all characters in the given string from
 * position 'i' onward to the left, and fill in
 * final character with '\0'
 */
void shift(char * s, const int i) {
  if (i < 0) return;

  if (DEBUG) {
	printf(" * Shifting characters for string '%s' at position '%d', length: '%d'\n", s, i, (int)strlen(s));
  }

  const int n = strlen(s);
  for (int j = i + 1; j < n; j++) {
	s[j - 1] = s[j];
  }
  s[n - 1] = '\0';

  if (DEBUG) {
	printf(" * String after: '%s', length: '%d'\n", s, (int)strlen(s));
  }
}


/**
 * Remove duplicate characters from a string 's'
 * TODO: Add error-checking for realloc call
 */
char * rmdup(char * s) {
  int n = strlen(s);
  int i = 0, j = 0;
  while (i < n) {
	j = i + 1;
	while (j < n) {
	  if (s[i] == s[j]) {
		shift(s, j);
		n--;
	  } else {
		j++;
	  }
	}
	i++;
  }

  s = realloc(s, sizeof(char) * n + 1);
  s[n] = '\0';
  return s;
}


/**
 * Run a test case
 * TODO: Handle possible errors from str_alloc and rmdup
 */
void test(const char * input, const char * expected) {
  const int n = strlen(input);
  // char * input_copy = (char *)malloc(sizeof(char) * n + 1);
  char * input_copy = str_alloc(n);
  strncpy(input_copy, input, n);
  input_copy = rmdup(input_copy);
  const int m = strlen(input_copy);

  if (strncmp(input_copy, expected, m) != 0) {
	printf("[FAIL]    Test case for input '%s' failed:\n", input);
  } else {
	printf("[SUCCESS] Test case for input '%s' passed!\n", input);
  }

  printf(" * Expected: '%s', length: '%d'\n", expected, (int)strlen(expected));
  printf(" * Actual:   '%s', length: '%d'\n", input_copy, (int)strlen(input_copy));
  printf("\n");
  free(input_copy);
}


int main() {

  printf("Running tests...\n\n");
  test("", "");
  test("a", "a");
  test("asdf", "asdf");
  test("asdfasdf", "asdf");
  test("asdfghjklqwertyuiopa", "asdfghjklqwertyuiop");
  test("aaaaaaaaaaaaaaaaaaaaaaaaaaa", "a");

  return 0;
}

